proposal: `Embedders` design (#5390)
deepset-ai/haystack: proposals/text/5390-embedders.md

- Retrievers should't be responsible for embedding Documents.
- Currently, Retrievers have many parameters just to support and configure different underlying Encoders(≈Embedders).
- Adding support for new embedding providers or strategies is difficult. It requires changing the Retriever code.

# Detailed design

## Handle queries and Documents
This is the most critical aspect of the design.

- When embedding queries, we expect the Embedder to transform a list of string into a list of vectors.
- When embedding Documents, we expect the Embedder to enrich a list of Documents with the corresponding vectors (stored in the `embedding` field).
- For Documents, we may want to embed some meta fields in addition to the content. Therefore, the Embedder should also handle this preparation work, which involves joining all the relevant content into a string to be embedded.

There have been much discussion about this point.
@ZanSara formulated the following implementation idea, that I like.

We make three classes:

- A `BasicEmbedder`, which is NOT a component, handling raw data + a factory method to reuse instances
```python
class HFBasicEmbedder:
    """

>     NOT A COMPONENT!

Reviewer: If we provide all the requirements to implement an embedder (e.g. being a singleton) through the `@embedder` decorator, I think we can just assume `HFBasicEmbedder` does whatever `HFTextEmbedder` is supposed to do, sparing one class. `HFDocumentEmbedder` would still derive from `HFTextEmbedder` so it inherits all the "embedder" capabilities

query_pipe.connect("embedder", "retriever")
query_pipe.connect("retriever", "reader")

results = query_pipe.run(...)
```

- The `OpenAITextEmbedder` uses OpenAI models to convert a list of strings into a list of vectors. It is used in the query pipeline to embed the query.
- The `OpenAIDocumentEmbedder` uses OpenAI models to enrich a list of Documents with the corresponding vectors (stored in the `embedding` field). It is used in the indexing pipeline to embed the Documents.
- The Retriever is no longer needed in the indexing pipeline.

# Motivation

The motivations behind this change were already provided in the previous proposals ([Embedding Retriever](3558-embedding_retriever.md) and [DocumentStores and Retrievers](4370-document_stores_and_retrievers.md)). Here is a summary:
- Retrievers should't be responsible for embedding Documents.
- Currently, Retrievers have many parameters just to support and configure different underlying Encoders(≈Embedders).
- Adding support for new embedding providers or strategies is difficult. It requires changing the Retriever code.

# Detailed design

## Handle queries and Documents
This is the most critical aspect of the design.

> - When embedding queries, we expect the Embedder to transform a list of string into a list of vectors.

Reviewer: ```suggestion
- When embedding queries, the Embedder component receives a list of strings in input that are transformed into a list of vectors returned as output.
```

> - When embedding Documents, we expect the Embedder to enrich a list of Documents with the corresponding vectors (stored in the `embedding` field).

Reviewer: ```suggestion
- When embedding documents, the Embedder component receives a list of `Document` objects in input; for each item in the list, the corresponding vectors are computed and stored in the `embedding` field of the item itself. The list is then returned as the component output.
```

> - For Documents, we may want to embed some meta fields in addition to the content. Therefore, the Embedder should also handle this preparation work, which involves joining all the relevant content into a string to be embedded.

Reviewer: ```suggestion
- When working with documents, there's the possibility to compute embeddings also for document's metadata. In this case, the Embedder will be responsible for performing any text-manipulation work needed in preparation of the actual embedding process.
```

**Below, I will focus on the public API. The internal implementation is discussed in [Implementation details](#implementation-details).**

```python
@component
class HFTextEmbedder:

>     class Input:

Reviewer: can we use the new API here?
```py
class HFTextEmbedder:

    @component.output_types(result=List[np.ndarray])
    ref run(self, data: List[str])
        return self.output(self.embedding_service.embed(data.data))
```

        data: List[str]

    class Output
        embeddings: List[np.ndarray]

    ...

    def run(self, data):

>         return self.output(self.embedding_service.embed(data.data))

Reviewer: ```suggestion
        return self.output(list_of_computed_embeddings)
```

@component
class HFDocumentEmbedder(HFTextEmbedder):

    class Input:
        documents: List[Document]

    class Output
        documents: List[np.ndarray]

    def run(self, data):

>         text_strings = [document.content for document in data.documents]
>         embeddings = self.embedding_service.embed(text_strings)
>         documents_with_embeddings = [Document.from_dict(**doc.to_dict, "embedding": emb) for doc, emb in zip(documents, embeddings)]
>         return self.output(documents = documents_with_embeddings)

Reviewer: ```suggestion
        return self.output(list_of_documents_with_embeddings)
```

```

## Different providers/strategies

> - We can define different classes depending on the providers: `OpenAIEmbedder`, `CohereEmbedder`, `HuggingFaceEmbedder`, `SentenceTransformersEmbedder`, etc.

Reviewer: ```suggestion
- We can define different embedder components depending on the models or services providing the actual embeddings: `OpenAIEmbedder`, `CohereEmbedder`, `HuggingFaceEmbedder`, `SentenceTransformersEmbedder`, etc.
```

Some examples:
- In Dense Passage Retrieval, you need a model to encode queries and another model to encode Documents
- in the TableTextRetriever, we use 3 different models: one for queries, one for textual passages and one for tables
- in Multimodal Retrieval, we can specify different models to encode queries and Documents

Since the Embedder will not be included in the Retriever, it makes sense to have different Embedders, each one using a single model.

```python
dpr_query_embedder = SentenceTransformersTextEmbedder(model_name="facebook/dpr-question_encoder-single-nq-base")
dpr_doc_embedder = SentenceTransformersDocumentEmbedder(model_name="facebook/dpr-ctx_encoder-single-nq-base")
```

## Implementation details

*You can skip this section if you are primarily interested in user experience.*

There have been much discussion on how to effectively implement this proposal.
The most important aspects to consider:
- we want different Embedders for queries and Documents as they require a different treatment
- if the same model is internally used for different Embedders, we want to reuse the same instance in order to save memory

> @ZanSara formulated the following implementation idea, that I like.

Reviewer: Nit: being a proposal, we should avoid details about its evolution, who did what and when, what we likes or don't - let's focus on explaining the technology.

> We make three classes:

Reviewer: ```suggestion
On top of the embedder components we already discussed, we introduce one additional abstraction:
```

> - An `EmbeddingService`, which is NOT a component, handling raw data + a factory method to reuse instances. It will live in a different package, **is for internal use only and will never be user-facing.** (The name of this class can be changed; feel free to propose better alternatives)

Reviewer: ```suggestion
- An `EmbeddingService`, which is NOT a component, responsible for performing the actual embedding computation, implemented as a singleton class in order to reuse instances. It will live in a different package and will be hidden from the public API.

```python

> class HFEmbeddingService:

Reviewer: ```suggestion
@singleton  # implementation is out of scope
class HFEmbeddingService:
```

    """
    NOT A COMPONENT!
    """

>     instances: List[HFEmbeddingService] = []
> 
>     def __new__(cls, *args, **kwargs):
> 		"""
>         Factory method.
> 		If an instance with the same identical params was already created,
>         return that instance instead of initializing a new one.
>         """
>         if <hash of name and init params> in HFEmbeddingService.instances:
>           return HFEmbeddingService.instances[<hash of name and init params>]
> 
>         embedding_service = cls(*args, **kwargs)
>         HFEmbeddingService.instances[<hash of name and init params>] = embedding_service
>         return embedding_service

Reviewer: ```suggestion
```

    def __init__(self, model_name: str, ... init params ...):
        """
        init takes the minimum parameters needed at init time, not
        the params needed at inference, so they're easier to reuse.
        """
        self.model = ...

    def embed(self, data: str, ... inference params ... ) -> np.ndarray:
        # compute embedding
        return embedding

class OpenAIEmbeddingService:
    ... same as above ...
```

> Given that EmbeddingServices are created through a factory method, when you request an instance, if another identical exists, the method returns that instance instead of a new one.
> 
> This makes model reusability automatic in all cases, which can save lots of memory without asking the user to think about it.

Reviewer: ```suggestion
Implemented as singletons, when instantiating an EmbeddingService class, if another identical one exists, the existing one will be returned without allocating additional resources for a new one. This makes model reusability transparent, saving lots of memory without any user intervention.
```

> - A `(Text/Table/Image/Audio)Embedder` component that does nothing but “wrapping” an EmbeddingService.

Reviewer: ```suggestion
This is how an EmbeddingService would be used by a text embedder component:

**Part of the public API**.
```python
@component
class HFTextEmbedder:

    class Input:
        data: List[str]

    class Output
        embeddings: List[np.ndarray]

    def __init__(self, model_name: str, ... init params ...):
        self.model_name = model_name
        self.model_params = ... params ...

    def warm_up(self):
        self.embedding_service = HFEmbeddingService(self.model_name, **self.model_params)

    def run(self, data):
        return self.output(self.embedding_service.embed(data.data))
```

> - A DocumentEmbedder component that handles the documents and the offloads the computation to an embedder.

Reviewer: ```suggestion
Another example, using an embedder component expecting documents:
```

**Part of the public API**.

```python
@component

> class HFDocumentEmbedder(HFTextEmbedder):

Reviewer: ```suggestion
class HFDocumentEmbedder:
```

  """
  Note: in this toy example inheritance from HFTextEmbedder makes sense because
  init and warm_up are identical. If they would differ significantly, let's remove
  the inheritance to simplify the architecture: it's not mandatory for the rest
  of the system to work.
  """

    class Input:
        documents: List[Document]

    class Output
        documents: List[np.ndarray]

    def run(self, data):
        text_strings = [document.content for document in data.documents]
        embeddings = self.embedding_service.embed(text_strings)
        documents_with_embeddings = [Document.from_dict(**doc.to_dict, "embedding": emb) for doc, emb in zip(documents, embeddings)]
        return self.output(documents = documents_with_embeddings)
```

# Drawbacks

> The drawbacks of concept separation between Retrievers and Embedders were discussed in [this proposal](https://github.com/deepset-ai/haystack/blob/main/proposals/text/4370-documentstores-and-retrievers.md) and mainly consist of **migration effort**.

Reviewer: ```suggestion
The drawbacks of separating Retrievers and Embedders were already discussed in [this proposal](https://github.com/deepset-ai/haystack/blob/main/proposals/text/4370-documentstores-and-retrievers.md) and mainly consist of **migration effort**.
```

> Several ideas were discussed (mostly concerning how to handle queries and Documents in the Embedders) and this seems to be the best approach. However, there are some potential cons to consider:

Reviewer: ```suggestion
Regarding the design proposed in this document, there are some potential drawbacks to consider:
```

  - Proliferation of classes (though they will be small and easy to maintain).

>   - Users need to properly understand which models are appropriate for a task or for embedding queries vs. Documents (see [Different models in the same embedding/retrieval task](#different-models-in-the-same-embeddingretrieval-task)). On the other hand, this can help raise their awareness.

Reviewer: ```suggestion
  - Users need to know which models are appropriate for which task (e.g. embedding queries rather than embedding documents, see [Different models in the same embedding/retrieval task](#different-models-in-the-same-embeddingretrieval-task)). On the other hand, this approach is more explicit and will help making users aware of problems and tradeoffs related to the topic.
```
